context.go

```diff
@@ -0,0 +1,196 @@
+package inject
+
+import (
+	"reflect"
+	"sort"
+	"strings"
+)
+
+// ctx is the dependency resolution context. It is used to detect circular
+// dependencies and provide the dependency tree.
+type ctx struct {
+	root    *stack
+	current *stack
+}
+
+func zeroCtx() ctx {
+	return ctx{}
+}
+
+func newCtx(inj *injector) ctx {
+	itype := "root"
+	if inj.parent != nil {
+		itype = "child"
+	}
+	root := newStack(nil, rootBindingKey{itype}, rootBinding{inj.name})
+	return ctx{root: root, current: root}
+}
+
+func (c *ctx) push(key bindingKey, binding resolvedBinding) (err error) {
+	if c.root == nil {
+		return nil
+	}
```

Comment:
I don't see a place where c.root can be nil

Reply:
zeroCtx()